test/test_foreach.py

```diff
@@ -646,38 +591,33 @@ def test_binary_op_tensors_on_different_devices(self, device, dtype, op):
         else:
             self.assertEqual(actual, tensors1)
 
-    @dtypes(*torch.testing.get_all_dtypes(include_bfloat16=True))
-    def test_pointwise_op_tensors_on_different_devices(self, device, dtype):
-        if self.device_type != 'cuda':
-            self.skipTest('CUDA is necessary for tests with tensors on different devices')
-
-        pointwise_ops = [
-            (torch._foreach_addcmul, torch._foreach_addcmul_, torch.addcmul),
-            (torch._foreach_addcdiv, torch._foreach_addcdiv_, torch.addcdiv),
-        ]
-        for foreach_op, foreach_op_, native_op in pointwise_ops:
-            # tensors1: ['cuda', 'cpu]
-            # tensors2: ['cuda', 'cpu]
-            # tensors3: ['cuda', 'cpu]
-            _cuda_tensors = self._get_test_data(device, dtype, 3)
-            _cpu_tensors = self._get_test_data('cpu', dtype, 3)
-            tensors1, tensors2, tensors3 = list(tensors for tensors in zip(_cuda_tensors, _cpu_tensors))
+    @onlyCUDA
+    @ops(foreach_pointwise_op_db)
+    def test_pointwise_op_tensors_on_different_devices(self, device, dtype, op):
+        # tensors1: ['cuda', 'cpu]
+        # tensors2: ['cuda', 'cpu]
+        # tensors3: ['cuda', 'cpu]
+        _cuda_tensors = op.sample_inputs(device, dtype, 3, same_size=True)
+        _cpu_tensors = op.sample_inputs('cpu', dtype, 3, same_size=True)
+        tensors1, tensors2, tensors3 = list(tensors for tensors in zip(_cuda_tensors, _cpu_tensors))
 
-            try:
-                actual = foreach_op(tensors1, tensors2, tensors3)
-            except RuntimeError as e:
-                with self.assertRaisesRegex(type(e), re.escape(str(e))):
-                    expected = [native_op(t1, t2, t3) for t1, t2, t3 in zip(tensors1, tensors2, tensors3)]
-            else:
+        foreach_op, foreach_op_ = op.method_variant, op.inplace_variant
+        native_op, native_op_ = op.ref, op.ref_inplace
+        try:
+            actual = foreach_op(tensors1, tensors2, tensors3)
+        except RuntimeError as e:
+            with self.assertRaisesRegex(type(e), re.escape(str(e))):
```

Comment:
when do errors happen here? It seems like inputs corresponding to the same op are always on the same device, e.g. the below translates into 
```
[native_op(*_cuda_tensors), native_op(*_cpu_tensors)]
```
which should work. Also, no harm in writing it like this, it's a bit cleaner than zip of previously zipped tensors.

Reply:
> when do errors happen here?

cpu kernels don't support bfloat16 and half, but @ops(foreach_pointwiese_op_db) creates the test cases of these two dtypes. Do you think using `@dtypes` decorator and limiting the used dtypes to, e.g., all_fp_dtypes?

> It seems like inputs corresponding to the same op are always on the same device,...

that sounds way better, thank you.

Reply:
Yeah, if the purpose of this test is to test behavior on the different devices, it's better to do it for a single datatype (e.g. float). But you probably want to actually test that inputs on the different devices through an error, and now you are not doing that.